micrometer-core/src/test/java/io/micrometer/core/instrument/binder/grpc/GrpcAsyncTest.java

```diff
@@ -119,6 +124,76 @@ void simulate_trace_in_async_requests() {
         });
     }
 
+    @Test
+    void multi_thread_client() {
+        AtomicReference<Observation> onStart = new AtomicReference<>();
+        AtomicReference<Observation> onMessage = new AtomicReference<>();
+        AtomicReference<Observation> halfClose = new AtomicReference<>();
+        ClientInterceptor clientInterceptor = new ClientInterceptor() {
+            @Override
+            public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
+                    CallOptions callOptions, Channel next) {
+                return new SimpleForwardingClientCall<>(next.newCall(method, callOptions)) {
+
+                    @Override
+                    public void start(Listener<RespT> responseListener, Metadata headers) {
+                        onStart.set(observationRegistry.getCurrentObservation());
+                        super.start(responseListener, headers);
+                    }
+
+                    @Override
+                    public void sendMessage(ReqT message) {
+                        onMessage.set(observationRegistry.getCurrentObservation());
+                        super.sendMessage(message);
+                    }
+
+                    @Override
+                    public void halfClose() {
+                        halfClose.set(observationRegistry.getCurrentObservation());
+                        super.halfClose();
+                    }
+                };
+            }
+        };
+
+        this.observationRegistry.observationConfig().observationHandler(context -> true);
+        this.channel = InProcessChannelBuilder.forName("sample")
+            .intercept(clientInterceptor)
+            .intercept(new ObservationGrpcClientInterceptor(this.observationRegistry))
+            .build();
+
+        SimpleServiceStub asyncStub = SimpleServiceGrpc.newStub(this.channel);
+        StreamObserver<SimpleResponse> responseObserver = mock(StreamObserver.class);
+        SimpleRequest request = SimpleRequest.newBuilder().setRequestMessage("Hello").build();
+
+        Observation ob = Observation.start("foo", this.observationRegistry);
+        ob.observe(() -> {
+            StreamObserver<SimpleRequest> requestObserver = asyncStub.clientStreamingRpc(responseObserver);
+
+            // perform onNext on a separate thread
+            AtomicBoolean onNextDone = new AtomicBoolean();
+            new Thread(() -> {
+                requestObserver.onNext(request);
+                onNextDone.set(true);
+            }).start();
+            await().untilTrue(onNextDone);
```

Comment:
I'm thinking if it is better to use a `CountDownLatch` instead of Awaitility for this.
```java
CountDownLatch onNextDone = new CountDownLatch();
new Thread(() -> {
    requestObserver.onNext(request);
    onNextDone.countDown();
    }).start();
onNextDone.await();
```

Also, maybe using `CompletableFuture` (uses the common pool) instead of creating a new `Thred` might be better.